kentik_api_library/kentik_api/synthetics/api_transport_grpc.py

```diff
@@ -0,0 +1,880 @@
+    def __init__(
+        self, credentials: Tuple[str, str], url: str = "synthetics.api.kentik.com:443", proxy: Optional[str] = None
+    ) -> None:
+        (email, token) = credentials
+        self._url = url
+        self._admin = SyntheticsAdminService()
+        self._data = SyntheticsDataService()
+        self._credentials = [
+            ("x-ch-auth-email", email),
+            ("x-ch-auth-api-token", token),
+        ]
+
+    def req(self, op: str, **kwargs) -> Any:
+        OPS: Dict[str, Callable] = {
+            "TestsList": self.tests_list,
+            "TestCreate": self.test_create,
+            "TestGet": self.test_get,
+            "TestPatch": self.test_patch,
+            "TestDelete": self.test_delete,
+            "TestStatusUpdate": self.test_status_update,
+            "AgentsList": self.agent_list,
+            "AgentGet": self.agent_get,
+            "AgentPatch": self.agent_patch,
+            "AgentDelete": self.agent_delete,
+            "GetHealthForTests": self.get_health_for_tests,
+            "GetTraceForTest": self.get_trace_for_test,
+        }
+
+        try:
+            svc = OPS[op]
+        except KeyError:
+            raise RuntimeError(f"Invalid operation '{op}'")
+
+        try:
+            return svc(**kwargs)
+        except RpcError as error:
+            raise new_api_error(error) from error
+
+    def tests_list(self, **kwargs) -> List[SynTest]:
+        tests: List[SynTest] = []
+        pb_tests = self._admin.ListTests(ListTestsRequest(), metadata=self._credentials, target=self._url).tests
+        for pbt in pb_tests:
+            if pbt.type in [TestType.none.value, TestType.bgp_monitor.value]:
+                pb_test = SynTest("[empty]")
+                populate_test_from_pb(pbt, pb_test)
+                tests.append(pb_test)
+            if pbt.type == TestType.mesh.value:
+                pb_test = MeshTest("[empty]")
+                populate_mesh_test_from_pb(pbt, pb_test)
+                tests.append(pb_test)
+            else:
+                log.warning('Skipping test "%s" of unsupported type "%s"', pbt.name, pbt.type)
+        return tests
+
+    def test_create(self, **kwargs) -> SynTest:
+        test: SynTest = kwargs["test"]
+        test._id = ID("")  # TestCreate doesn't accept id
+        pb_test = pb_from_test(test)
+        response = self._admin.CreateTest(CreateTestRequest(test=pb_test), metadata=self._credentials, target=self._url)
+        out = SynTest("[empty]")
+        populate_test_from_pb(response.test, out)
+        return out
+
+    def test_get(self, **kwargs) -> SynTest:
+        id = str(kwargs["id"])
+        response = self._admin.GetTest(GetTestRequest(id=id), metadata=self._credentials, target=self._url)
+        out = SynTest("[empty]")
+        populate_test_from_pb(response.test, out)
+        return out
+
+    def test_patch(self, **kwargs) -> SynTest:
+        pb_test = pb_from_test(kwargs["test"])
```

Comment:
A  more user-friendly solution is to:
1) fetch the existing test (fail if it does not exist)
2) use (modified) `compare_tests(existing_test, new_test)` (or `compare_dict`) function to figure out the mask
3) call the `PatchTest` method.

Note that support for PATCH will be replaced by UPDATE.

Also, I've modified the `compare_dict` function in the [beta2 branch](https://github.com/kentik/synth_tools/blob/5395f5344ded7bae2cd63fffedfa4cd15287c336/kentik_synth_client/utils.py#L6) of synth_tools and use it in [compare_test](https://github.com/kentik/synth_tools/blob/5395f5344ded7bae2cd63fffedfa4cd15287c336/synth_tools/commands/tests.py#L253) command. It should be pretty usable for generating the `mask`.

Reply:
Will update while migrating to API 2022

Reply:
mask is no longer in the api for v202202